### PR TITLE
getUploadWidgetOptions, generateSignatureCallback, generateUploadWidgetResultCallback

### LINKED ISSUE #164: getUploadWidgetOptions
### Discussed in https://github.com/cloudinary-community/cloudinary-util/discussions/124

<div type='discussions-op-text'>

<sup>Originally posted by **colbyfayock** January 12, 2024</sup>
The [Cloudinary Upload Widget](https://cloudinary.com/documentation/upload_widget) provides developers a way to easily drop in a clientside upload experience including drag and drop, the ability to select from multiple sources, and ways to easily configure upload controls. The CldUploadWidget component is a wrapper around this widget giving an easier way to set up and configure the widget in framework applications ([Next.js](https://next.cloudinary.dev/clduploadwidget/basic-usage), [Nuxt](https://cloudinary.nuxtjs.org/components/clduploadwidget), [Svelte](https://svelte.cloudinary.dev/clduploadwidget/usage/)).

CldUploadWidget currently utilizes the Script installation method, which is a bit beyond the scope of this, but ultimately it requires a group of options to be passed into it in order to initialize and make use of the wide variety of features available to the Upload Widget.

In order to provide as much parity as possible between the libraries maintaining a CldUploadWidget component, the maintainers and contributors must rewrite or copy and paste similar code within each of their frameworks, which is slow and prone to creating implementation differences. To help alleviate this, we can abstract some of the work to a utility library, much like we're currently doing for building URLs with CldImage.

## The Challenge

Unlike CldImage, CldUploadWidget doesn't take a URL as the primary input, and instead takes a configuration object which is composed of the different options available. This means there's not a single interface that can compose the entirety of CldUploadWidget.

Additionally, some of the initialization and work needed to manage the component is different depending on the framework, leaving no point to try to deduplicate that work.

## Proposed Solution

While we can't fully abstract the component into a single function, we can think about how we can compose different parts in a reusable fashion.

This proposal suggests to abstract out the options that are passed into the player upon initialization as well as the Types that are used, both from the Cloudinary Upload Widget itself, but also the props that are passed in as part of the options configuration.

### getUploadWidgetOptions

The getUploadWidgetOptions function will compose the options that need to be passed into the Cloudinary Upload Widget createUploadWidget method upon initialization. 

Usage would look like the following:

```
const options = getUploadWidgetOptions(props);
cloudinary.createUploadWidget(ref, options)
```

This will simplify the ability to support all Upload Widget features while looking forward to creating different API surfaces for a better Upload Widget developer experience.

### Types

There are two Types that are needed to be maintained for the Upload Widget:
* Cloudinary Upload Widget (which are not publicly available)
* getUploadWidgetOptions

**Cloudinary Upload Widget** types can be maintained in a new utility package of @cloudinary-util such as @cloudinary-util/types, which will make it easy to import these types into any project, though with this abstract, may possibly be no longer needed in the framework layer.

**getUploadWidgetOptions** will be the primary interface that the frameworks interact with, which can be used as the foundational interface that creates the component's Types, for instance

```
interface CldUploadWidget extends GetUploadWidgetOptions {
   …
}
```

## Relevant Issues

* https://github.com/colbyfayock/cloudinary-util/issues/122

## Other Similar RFCs
* https://github.com/colbyfayock/cloudinary-util/discussions/86</div>

### PR DESCRIPTION
**Pull request by colbyfayock**
# Description

This creates a new set of helpers that will help framework libraries reduce the boilerplate needed to instantiate new instances of the upload widget.

* getUploadWidgetOptions: this takes in a set of options that deal with the configuration of the upload widget, including account credentials and signature callback function
* generateSignatureCallback: this generates a new callback function that given an endpoint and fetcher (like the fetch API), will make it easier to pass the resulting function to the widget options above
* generateUploadWidgetResultCallback: when uploading with the widget, different events are produced depending on the state of the upload, this helps manage all of the different events via a single callback function with the ability to pass in and tap into the different events

## Issue Ticket Number

Fixes #164 

## Usage

```
const signatureCallback = generateSignatureCallback({
  signatureEndpoint: '/asdf',
  fetch
})

const uploadOptions = getUploadWidgetOptions({
  signatureCallback
}, {
  cloud: {
    cloudName: 'testcloud',
    apiKey: 'abcd1234'
  }
});

const resultsCallback = generateUploadWidgetResultCallback({
  onError: (uploadError) => {
    // Example on storing error
    // setError(uploadError);
  },
  onResult: (uploadResult) => {
    // Example of storing results on any event
    // setResults(uploadResult);
  },
  onSuccess: (uploadResult) => {
    // Tapping into individual events
  }
});

cloudinary.createUploadWidget(uploadOptions, resultsCallback);
```

### CODE REVIEW

File: packages/url-loader/src/lib/upload.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { CloudinaryUploadWidgetOptions } from '@cloudinary-util/types';
+
+/**
+ * generateSignature
+ * @description Makes a request to an endpoint to sign Cloudinary parameters as part of widget creation
+ */
+
+export interface GenerateSignatureCallback {
+  fetch: Function;
```
Comment (ghostdevv) [inline packages/url-loader/src/lib/upload.ts:9]: might be able to use `typeof fetch` here which should be globally available

**Review by ghostdevv — APPROVED** [review]

File: packages/url-loader/src/lib/upload-widget.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  'show-completed': 'onShowCompleted',
+  'source-changed': 'onSourceChanged',
+  'success': 'onSuccess',
+  'tags': 'onTags',
+  'upload-added': 'onUploadAdded',
+}
+
+export function generateUploadWidgetResultCallback(options: GenerateUploadWidgetResultCallback) {
+  return function resultCallback(error: CloudinaryUploadWidgetError, uploadResult: CloudinaryUploadWidgetResults) {
+    if ( error && error !== null ) {
```
Comment (Baroshem) [inline packages/url-loader/src/lib/upload-widget.ts:97]: question: is this full statement necessary? AFAIK, null is falsy so if error is truthy then it means automatically that it cannot be null (falsy)

**Review by Baroshem — APPROVED** [review]
One small question, apart from that Approve :)

File: packages/url-loader/src/lib/upload-widget.ts (hunk 2 above)
Comment (ghostdevv) [reply packages/url-loader/src/lib/upload-widget.ts:97]: null is falsy yea

Comment (colbyfayock) [reply packages/url-loader/src/lib/upload-widget.ts:97]: sounds good, was just pasting this over from existing code, will clean that up!

File: packages/url-loader/src/lib/upload.ts (hunk 1 above)
Comment (colbyfayock) [reply packages/url-loader/src/lib/upload.ts:9]: good point, though my original thinking was someone could pass in something that's _not_ fetch. realistically its going to be fetch and the shape supports fetch "as is", so perhaps i just move forward with the suggestion and can look at other use cases should they ever come up

**Pull request merged**